param.go

```diff
@@ -129,16 +137,111 @@ func newParamList(ctype reflect.Type) (paramList, error) {
 		}
 		pl.Params = append(pl.Params, p)
 	}
+
 	return pl, nil
 }
 
+func (pl paramList) Build(*Container) (reflect.Value, error) {
+	panic("paramList.Build() must never be called")
+}
+
+// BuildParams returns an ordered list of values which may be passed directly
+// to the underlying constructor.
+func (pl paramList) BuildParams(c *Container) ([]reflect.Value, error) {
+	args := make([]reflect.Value, len(pl.Params))
+	for i, p := range pl.Params {
+		var err error
+		args[i], err = p.Build(c)
+		if err != nil {
+			return nil, errWrapf(err, "could not build argument %d for constructor %v", i, pl.ctype)
+		}
+	}
+	return args, nil
+}
+
 // newParam builds a param from the given type. If the provided type is a
 // dig.In struct, an paramObject will be returned.
 func newParam(t reflect.Type) (param, error) {
-	if IsIn(t) {
+	switch {
+	case IsIn(t):
 		return newParamObject(t)
+	case embedsType(t, _inPtrType):
+		return nil, fmt.Errorf(
+			"%v embeds *dig.In which is not supported, embed dig.In value instead", t)
+	case t.Kind() == reflect.Ptr && IsIn(t.Elem()):
+		return nil, fmt.Errorf(
+			"dependency %v is a pointer to dig.In, use value type instead", t)
```

Comment:
I realize that you've just copied the existing error message, but this wording is really confusing to me - there's no pointer to `dig.In` involved here at all. Would this be better as `cannot depend on pointer to params struct, use a value instead`?